### PR TITLE
fix: encode Fork neighbour prefix one nibble per byte (odd-length panic)

### PR DESCRIPTION
**Pull request by mgpai22**
## Summary

A `Fork` proof step encoded the neighbour prefix with `nibblesToBytes`, which packs two nibbles per byte and indexes `data[i+1]` — this **panics on odd-length prefixes** (`index out of range`), and even for non-empty even-length prefixes it produces bytes the on-chain Aiken validator rejects.

The Aiken `merkle-patricia-forestry` validator consumes a `Fork` neighbour prefix **nibble-by-nibble** (`combine(neighbor.prefix, neighbor.root)`), and the reference `@aiken-lang/merkle-patricia-forestry` library serialises it as `nibbles(prefix)` — i.e. **one nibble per byte** (each nibble value 0–15 as a full byte). This PR encodes the `Fork` prefix that way so proofs with odd- (and even-) length fork prefixes serialise correctly and verify on-chain.

## Changes
- `nibble.go`: add `nibblesToExpandedBytes` (one nibble per byte, any length); make `nibblesToBytes` odd-safe (trailing nibble → low half) as a defensive measure.
- `proof.go`: `ProofStep.MarshalCBOR` `Fork` case uses `nibblesToExpandedBytes(s.neighbor.prefix)`. (The `Leaf` `key` is the full 32-byte hash — always 64 nibbles/even — and is consumed packed, so it is unchanged.)
- `proof_oddfork_test.go`: golden tests for 1/2/3-nibble fork prefixes whose CBOR matches the reference TS library byte-for-byte, plus encoder unit tests.

## Validation
Reproduced against a live preprod Aiken MPF validator: before, building a withdrawal whose proof contains an odd-length fork prefix panics in `nibblesToBytes`; after, the withdrawal builds, the redeemer's fork prefixes decode correctly, and the transaction submits and **confirms on-chain** (the validator's `combine` accepts the proof). Multiple such withdrawals confirmed across consecutive blocks.

---
## Summary by cubic
Fixes odd-length panics and wrong encoding for `Fork` neighbour prefixes by encoding one nibble per byte to match the on-chain Aiken validator. Adds `Trie.ProveInsert` for insert/update proofs when the key is absent. Updates CI to Go 1.25.x so `nilaway@latest` installs.

- **Bug Fixes**
  - Encode `Fork` neighbour prefix nibble-by-nibble via `nibblesToExpandedBytes` in `ProofStep.MarshalCBOR`, matching Aiken `merkle-patricia-forestry` and `@aiken-lang/merkle-patricia-forestry`.
  - Make `nibblesToBytes` safe for odd inputs (trailing nibble in low half); `Leaf` key packing unchanged.
  - Add tests: encoder cases (empty/even/odd) and golden CBOR for 1/2/3-nibble prefixes.

- **New Features**
  - Add `Trie.ProveInsert` to build membership proofs for insert/update against the current root when the key is absent.

<sup>Written for commit d72554ebf304fe9c23b84fa6875a7dfe4725eaa2. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/blinklabs-io/merkle-patricia-forestry/pull/195?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed nibble-to-byte packing to correctly handle odd-length input sequences instead of dropping trailing data.

* **New Features**
  * Added encoding method for nibbles where each nibble occupies one byte in the output.

* **Tests**
  * Added test coverage for odd-length nibble handling and proof generation with edge cases.

### CODE REVIEW

File: proof.go [hunk 1]
```diff
@@ -151,7 +151,10 @@ func (s *ProofStep) MarshalCBOR() ([]byte, error) {
 		return cbor.Encode(tmpData)
 
 	case ProofStepTypeFork:
-		prefixBytes := nibblesToBytes(s.neighbor.prefix)
+		// The Fork neighbour prefix is consumed nibble-by-nibble on-chain
+		// (combine(neighbor.prefix, neighbor.root) in the Aiken validator),
+		// so it must be encoded one nibble per byte rather than packed.
+		prefixBytes := nibblesToExpandedBytes(s.neighbor.prefix)
```
Comment (wolf31o2) [inline proof.go:157]: Current main now defines `nibblesToIndividualBytes` together with `individualBytesToNibbles` for CBOR round trips. Please resolve this by reusing that helper instead of retaining a second equivalent `nibblesToExpandedBytes`; otherwise the merged code has duplicate implementations.

File: trie.go [hunk 2]
```diff
@@ -215,3 +215,89 @@ func (t *Trie) Prove(key []byte) (*Proof, error) {
...
+
+		childIdx := int(pathMinusPrefix[0])
+		leafChildIdx := int(leafSuffixMinusPrefix[0])
+		leafCopy := newLeaf(leafSuffixMinusPrefix[1:], n.key, n.value)
+
+		var children [16]Node
+		children[leafChildIdx] = leafCopy
+
+		proof := newProof(path, nil)
+		proof.Rewind(childIdx, len(prefix), children[:])
```
Comment (wolf31o2) [inline trie.go:251]: On current `main`, `nilaway ./...` traces this zero-value array through `Proof.Rewind` to the slices in `merkleProof` and fails the check. Please allocate it with `children := make([]Node, 16)` and pass `children` directly. The same change is needed for the second temporary child vector below.

File: trie.go [hunk 3]
```diff
@@ -215,3 +215,89 @@ func (t *Trie) Prove(key []byte) (*Proof, error) {
 	path := keyToPath(key)
 	return t.rootNode.generateProof(path)
 }
+
+// ProveInsert returns the proof required to insert or update the specified key
+// against the trie's current root.
+func (t *Trie) ProveInsert(key []byte) (*Proof, error) {
```
Comment (wolf31o2) [inline trie.go:221]: Please add regression coverage for this new public API. At minimum, compare its encoded proof with `Prove` after actually inserting the same key across an empty trie, a divergent root leaf or branch, an empty child, a nested branch, and an existing-key update; also assert that `ProveInsert` leaves the source root unchanged.

**Review by wolf31o2 — CHANGES_REQUESTED** [review]
Requesting changes before merge. Current main already contains the fork-prefix CBOR helper and its decoder, so the conflict should be resolved by retaining that implementation. The new ProveInsert path also fails the current NilAway check and needs regression coverage proving that its generated proof matches a real insertion without mutating the source trie.
